File: coin-change.java
/*
Did it run on leetcode: Yes
Time Complexity: 0(N*M)
Space Complexity: 0(N)
Problems faced: 
Initially started with 2-d matrix but did not work out.
required modifications from paper to 1-d matrix

Algorithm:
We create a `change` array equal to size of amount.

Lets say we have coins {1,2,5} amount=5
initially we will find how many ways we can form with coin-1.
Then we find in how many ways we can form amoun-5 combined of {1,2}.
Then we find in how many ways we can form amoun-5 combined of {1,2,3}.

Which will give us the answer.


*/



class Solution {
    public int change(int amount, int[] coins) {
        
        
        int[] ways = new int[amount+1];
        ways[0] = 1;
        for(int i=0;i<coins.length;++i){
            int coin = coins[i];
            for(int j=coin;j<=amount;++j){
                ways[j]+=ways[j-coin];
            }
        }
        
        return ways[amount];
    }
}

File: coinchange.swift
// Time Complexity : 0(amount*n)
// Space Complexity : 0(amount)
// Did this code successfully run on Leetcode :
// Any problem you faced while coding this :


// Your code here along with comments explaining your approach
    func change(_ amount: Int, _ coins: [Int]) -> Int {
        var dp = Array(repeating: Array(repeating: 0, count: amount + 1),count: 2)
        dp[0][0] = 1
        for i in 1...coins.count {
            for j in 0...amount {
                if coins[i-1] > j {
                    dp[1][j] = dp[0][j]
                } else {
                    dp[1][j] = dp[1][j - coins[i-1]] + dp[0][j]
                    dp[0][j] = dp[1][j]
                }
            }
        }
        return dp[1][amount]
    }